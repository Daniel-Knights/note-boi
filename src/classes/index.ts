export * from './appError';
export * from './debounceQueue';
export * from './encryptor';
export * from './fetchBuilder';
export * from './keyStore';
export * from './storage';
export * from './unsyncedNotesManager';
