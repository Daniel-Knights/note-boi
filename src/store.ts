import { reactive } from 'vue';
import { invoke } from '@tauri-apps/api/tauri';
import { v4 as uuidv4 } from 'uuid';

import { isEmptyNote } from './utils';

export class Note {
  readonly id = uuidv4();
  body = {
    delta: '',
    text: '',
  };
  timestamp = Date.now();
}

interface State {
  notes: Note[];
  selectedNote: Note;
}

export const state = reactive<State>({
  notes: [],
  selectedNote: new Note(),
});

/** Sorts notes in descending order by timestamp. */
function sortNotes() {
  state.notes.sort((a, b) => b.timestamp - a.timestamp);
}

/** Finds note index within {@link state.notes}. */
function findNoteIndex(id: string) {
  return state.notes.findIndex((nt) => nt.id === id);
}

/** Finds note within {@link state.notes}. */
export function findNote(id: string): Note | undefined {
  return state.notes.find((nt) => nt.id === id);
}

/** Deletes {@link state.selectedNote} when both `title` and `body` are empty. */
function clearEmptyNote(): void {
  const isValidClear = state.notes.length > 1;

  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote) return;

  if (isValidClear && isEmptyNote(foundNote)) {
    deleteNote(state.selectedNote.id);
  }
}

const selectEvent = new CustomEvent('note-select');

/**
 * Looks for note with given `id` in {@link state.notes},
 * and sets it to {@link state.selectedNote}.
 */
export function selectNote(id: string): void {
  if (state.selectedNote.id === id) return;

  clearEmptyNote();

  const foundNote = findNote(id);
  if (foundNote) state.selectedNote = { ...foundNote };

  document.dispatchEvent(selectEvent);
}

/** Fetches all notes and updates {@link state}. */
export async function getAllNotes(): Promise<void> {
  const fetchedNotes = await invoke<Note[]>('get_all_notes').catch(console.error);

  const hasNotes = fetchedNotes && fetchedNotes.length > 0;
  if (!hasNotes) return newNote();

  state.notes = fetchedNotes;

  sortNotes();

  if (isEmptyNote(fetchedNotes[0])) {
    state.notes[0].timestamp = Date.now();
  }

  state.selectedNote = { ...state.notes[0] };

  document.dispatchEvent(selectEvent);
}

/** Deletes note with the given `id`. */
export function deleteNote(id: string): void {
  state.notes.splice(findNoteIndex(id), 1);
  state.selectedNote = { ...state.notes[0] };

  sortNotes();

  invoke('delete_note', { id }).catch(console.error);
}

/** Creates an empty note. */
export function newNote(): void {
  const foundNote = findNote(state.selectedNote.id);

  // Only update timestamp if selected note is empty
  if (foundNote && isEmptyNote(foundNote)) {
    state.selectedNote.timestamp = Date.now();
    return;
  }

  const freshNote = new Note();

  state.notes.unshift(freshNote);
  state.selectedNote = { ...freshNote };

  invoke('new_note', { note: { ...freshNote } }).catch(console.error);

  document.dispatchEvent(selectEvent);
}

/** Edits note body on Quill `text-change`. */
export function editBody(delta: string, text: string): void {
  const foundNote = findNote(state.selectedNote.id);
  if (!foundNote || text === foundNote.body.text) return;

  const timestamp = Date.now();
  foundNote.timestamp = timestamp;
  state.selectedNote.timestamp = timestamp;

  foundNote.body = { delta, text };

  sortNotes();

  invoke('edit_note', { note: { ...foundNote } }).catch(console.error);
}
