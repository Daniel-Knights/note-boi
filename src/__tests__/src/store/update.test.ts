import { check } from '@tauri-apps/plugin-updater';

import * as u from '../../../store/update';
import { Storage } from '../../../classes';
import { clearMockApiResults, mockApi } from '../../mock';

describe('Update', () => {
  describe('handleUpdate', () => {
    it('Handles update', async () => {
      const { calls } = mockApi();

      await u.handleUpdate();

      assert.isTrue(u.updateState.isDownloading);
      assert.isTrue(u.updateState.isAvailable);
      assert.strictEqual(calls.size, 4);
      assert.isTrue(calls.tauriApi.has('plugin:updater|check'));
      assert.isTrue(calls.tauriApi.has('plugin:dialog|ask'));
      assert.isTrue(calls.tauriApi.has('plugin:updater|download_and_install'));
      assert.isTrue(calls.tauriApi.has('plugin:process|restart'));
    });

    it('Returns if update unavailable', async () => {
      const { calls, setResValues } = mockApi();

      setResValues.tauriApi({ checkUpdate: [null] });

      await u.handleUpdate();

      assert.isFalse(u.updateState.isDownloading);
      assert.isFalse(u.updateState.isAvailable);
      assert.strictEqual(calls.size, 1);
      assert.isTrue(calls.tauriApi.has('plugin:updater|check'));
    });

    it("Updates without asking, if updateStrategy is 'auto'", async () => {
      const { calls } = mockApi();

      u.setUpdateStrategy('auto');

      await u.handleUpdate();

      assert.isTrue(u.updateState.isDownloading);
      assert.isTrue(u.updateState.isAvailable);
      assert.strictEqual(calls.size, 3);
      assert.isTrue(calls.tauriApi.has('plugin:updater|check'));
      assert.isTrue(calls.tauriApi.has('plugin:updater|download_and_install'));
      assert.isTrue(calls.tauriApi.has('plugin:process|restart'));
    });

    it('Returns if version has been seen', async () => {
      const { calls } = mockApi();

      Storage.set('UPDATE_SEEN', '1.0.0');

      await u.handleUpdate();

      assert.isFalse(u.updateState.isDownloading);
      assert.isTrue(u.updateState.isAvailable);
      assert.strictEqual(calls.size, 1);
      assert.isTrue(calls.tauriApi.has('plugin:updater|check'));
    });

    it("Asks if user wants to update and sets seen version if they don't", async () => {
      const { calls, setResValues } = mockApi();

      setResValues.tauriApi({ askDialog: [false] });

      await u.handleUpdate();

      assert.isFalse(u.updateState.isDownloading);
      assert.isTrue(u.updateState.isAvailable);
      assert.strictEqual(calls.size, 2);
      assert.isTrue(calls.tauriApi.has('plugin:updater|check'));
      assert.isTrue(calls.tauriApi.has('plugin:dialog|ask'));
      assert.deepEqual(calls.tauriApi[1]!.calledWith, {
        message: 'A new version of NoteBoi is available.\nDo you want to update now?',
        title: 'Update available: v1.0.0',
        kind: undefined,
      });
      assert.strictEqual(Storage.get('UPDATE_SEEN'), '1.0.0');
    });
  });

  describe('updateAndRelaunch', () => {
    it('Updates and relaunches', async () => {
      const { calls } = mockApi();
      const mockUpdate = (await check())!;

      clearMockApiResults({ calls });

      await u.updateAndRelaunch(mockUpdate);

      assert.isTrue(u.updateState.isDownloading);
      assert.isFalse(u.updateState.isAvailable);
      assert.strictEqual(calls.size, 2);
      assert.isTrue(calls.tauriApi.has('plugin:updater|download_and_install'));
      assert.isTrue(calls.tauriApi.has('plugin:process|restart'));
    });

    it("Catches error and doesn't retry", async () => {
      const { calls, setErrorValue, setResValues } = mockApi();
      const mockUpdate = (await check())!;

      clearMockApiResults({ calls });
      setErrorValue.tauriApi('plugin:updater|download_and_install');
      setResValues.tauriApi({ askDialog: [false], downloadAndInstallUpdate: [] });

      await u.updateAndRelaunch(mockUpdate);

      assert.isFalse(u.updateState.isDownloading);
      assert.isFalse(u.updateState.isAvailable);
      assert.strictEqual(calls.size, 1);
      assert.isTrue(calls.tauriApi.has('plugin:dialog|ask'));
      assert.deepEqual(calls.tauriApi[0]!.calledWith, {
        message: 'Try again?',
        title: 'Unable to install update',
        kind: 'error',
      });
    });

    it('Catches error and retries', async () => {
      const { calls, setErrorValue, setResValues } = mockApi();
      const mockUpdate = (await check())!;

      clearMockApiResults({ calls });
      setErrorValue.tauriApi('plugin:updater|download_and_install');
      setResValues.tauriApi({ askDialog: [true, false] });

      await u.updateAndRelaunch(mockUpdate);

      assert.isFalse(u.updateState.isDownloading);
      assert.isFalse(u.updateState.isAvailable);
      assert.strictEqual(calls.size, 2);
      assert.isTrue(calls.tauriApi.has('plugin:dialog|ask', 2));
    });
  });

  it('setUpdateStrategy', () => {
    assert.strictEqual(u.updateState.strategy, 'manual');
    assert.isNull(Storage.get('UPDATE_STRATEGY'));

    u.setUpdateStrategy('auto');

    assert.strictEqual(u.updateState.strategy, 'auto');
    assert.strictEqual(Storage.get('UPDATE_STRATEGY'), 'auto');
  });
});
