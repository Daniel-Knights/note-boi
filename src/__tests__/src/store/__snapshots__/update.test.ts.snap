// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > Storage.get('THEME') 1`] = `null`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > Storage.get('UPDATE_SEEN') 1`] = `"1.0.0"`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > allCalls 1`] = `
[
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|check",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": {
        "kind": undefined,
        "message": "A new version of NoteBoi is available.
Do you want to update now?",
        "title": "Update available: v1.0.0",
      },
      "name": "plugin:dialog|ask",
    },
  ],
]
`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > openedPopup 1`] = `undefined`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > selectedTheme 1`] = `"System"`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > handleUpdate > Asks if user wants to update and sets seen version if they don't > updateState 1`] = `
{
  "isAvailable": true,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Update > handleUpdate > Handles update > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > handleUpdate > Handles update > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > handleUpdate > Handles update > Storage.get('THEME') 1`] = `null`;

exports[`Update > handleUpdate > Handles update > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Update > handleUpdate > Handles update > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Update > handleUpdate > Handles update > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > handleUpdate > Handles update > allCalls 1`] = `
[
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|check",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": {
        "kind": undefined,
        "message": "A new version of NoteBoi is available.
Do you want to update now?",
        "title": "Update available: v1.0.0",
      },
      "name": "plugin:dialog|ask",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|download_and_install",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:process|restart",
    },
  ],
]
`;

exports[`Update > handleUpdate > Handles update > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > handleUpdate > Handles update > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > handleUpdate > Handles update > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > handleUpdate > Handles update > openedPopup 1`] = `undefined`;

exports[`Update > handleUpdate > Handles update > selectedTheme 1`] = `"System"`;

exports[`Update > handleUpdate > Handles update > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > handleUpdate > Handles update > updateState 1`] = `
{
  "isAvailable": true,
  "isDownloading": true,
  "strategy": "manual",
}
`;

exports[`Update > handleUpdate > Returns if update unavailable > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > handleUpdate > Returns if update unavailable > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > handleUpdate > Returns if update unavailable > Storage.get('THEME') 1`] = `null`;

exports[`Update > handleUpdate > Returns if update unavailable > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Update > handleUpdate > Returns if update unavailable > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Update > handleUpdate > Returns if update unavailable > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > handleUpdate > Returns if update unavailable > allCalls 1`] = `
[
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|check",
    },
  ],
]
`;

exports[`Update > handleUpdate > Returns if update unavailable > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > handleUpdate > Returns if update unavailable > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > handleUpdate > Returns if update unavailable > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > handleUpdate > Returns if update unavailable > openedPopup 1`] = `undefined`;

exports[`Update > handleUpdate > Returns if update unavailable > selectedTheme 1`] = `"System"`;

exports[`Update > handleUpdate > Returns if update unavailable > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > handleUpdate > Returns if update unavailable > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Update > handleUpdate > Returns if version has been seen > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > handleUpdate > Returns if version has been seen > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > handleUpdate > Returns if version has been seen > Storage.get('THEME') 1`] = `null`;

exports[`Update > handleUpdate > Returns if version has been seen > Storage.get('UPDATE_SEEN') 1`] = `"1.0.0"`;

exports[`Update > handleUpdate > Returns if version has been seen > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Update > handleUpdate > Returns if version has been seen > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > handleUpdate > Returns if version has been seen > allCalls 1`] = `
[
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|check",
    },
  ],
]
`;

exports[`Update > handleUpdate > Returns if version has been seen > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > handleUpdate > Returns if version has been seen > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > handleUpdate > Returns if version has been seen > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > handleUpdate > Returns if version has been seen > openedPopup 1`] = `undefined`;

exports[`Update > handleUpdate > Returns if version has been seen > selectedTheme 1`] = `"System"`;

exports[`Update > handleUpdate > Returns if version has been seen > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > handleUpdate > Returns if version has been seen > updateState 1`] = `
{
  "isAvailable": true,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > Storage.get('THEME') 1`] = `null`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > Storage.get('UPDATE_STRATEGY') 1`] = `"auto"`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > allCalls 1`] = `
[
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|check",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|download_and_install",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:process|restart",
    },
  ],
]
`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > openedPopup 1`] = `undefined`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > selectedTheme 1`] = `"System"`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > handleUpdate > Updates without asking, if updateStrategy is 'auto' > updateState 1`] = `
{
  "isAvailable": true,
  "isDownloading": true,
  "strategy": "auto",
}
`;

exports[`Update > setUpdateStrategy > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > setUpdateStrategy > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > setUpdateStrategy > Storage.get('THEME') 1`] = `null`;

exports[`Update > setUpdateStrategy > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Update > setUpdateStrategy > Storage.get('UPDATE_STRATEGY') 1`] = `"auto"`;

exports[`Update > setUpdateStrategy > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > setUpdateStrategy > allCalls 1`] = `[]`;

exports[`Update > setUpdateStrategy > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > setUpdateStrategy > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > setUpdateStrategy > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > setUpdateStrategy > openedPopup 1`] = `undefined`;

exports[`Update > setUpdateStrategy > selectedTheme 1`] = `"System"`;

exports[`Update > setUpdateStrategy > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > setUpdateStrategy > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "auto",
}
`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > Storage.get('THEME') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > allCalls 1`] = `
[
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|check",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": {
        "kind": "error",
        "message": "Try again?",
        "title": "Unable to install update",
      },
      "name": "plugin:dialog|ask",
    },
  ],
]
`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > openedPopup 1`] = `undefined`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > selectedTheme 1`] = `"System"`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > updateAndRelaunch > Catches error and doesn't retry > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Update > updateAndRelaunch > Catches error and retries > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > updateAndRelaunch > Catches error and retries > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and retries > Storage.get('THEME') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and retries > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and retries > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and retries > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > updateAndRelaunch > Catches error and retries > allCalls 1`] = `
[
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|check",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": {
        "kind": "error",
        "message": "Try again?",
        "title": "Unable to install update",
      },
      "name": "plugin:dialog|ask",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": {
        "kind": "error",
        "message": "Try again?",
        "title": "Unable to install update",
      },
      "name": "plugin:dialog|ask",
    },
  ],
]
`;

exports[`Update > updateAndRelaunch > Catches error and retries > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > updateAndRelaunch > Catches error and retries > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > updateAndRelaunch > Catches error and retries > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > updateAndRelaunch > Catches error and retries > openedPopup 1`] = `undefined`;

exports[`Update > updateAndRelaunch > Catches error and retries > selectedTheme 1`] = `"System"`;

exports[`Update > updateAndRelaunch > Catches error and retries > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > updateAndRelaunch > Catches error and retries > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": false,
  "strategy": "manual",
}
`;

exports[`Update > updateAndRelaunch > Updates and relaunches > KeyStore.getKey() 1`] = `undefined`;

exports[`Update > updateAndRelaunch > Updates and relaunches > Storage.get('MENU_WIDTH') 1`] = `null`;

exports[`Update > updateAndRelaunch > Updates and relaunches > Storage.get('THEME') 1`] = `null`;

exports[`Update > updateAndRelaunch > Updates and relaunches > Storage.get('UPDATE_SEEN') 1`] = `null`;

exports[`Update > updateAndRelaunch > Updates and relaunches > Storage.get('UPDATE_STRATEGY') 1`] = `null`;

exports[`Update > updateAndRelaunch > Updates and relaunches > Storage.get('USERNAME') 1`] = `null`;

exports[`Update > updateAndRelaunch > Updates and relaunches > allCalls 1`] = `
[
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|check",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:updater|download_and_install",
    },
  ],
  [
    "tauriApi",
    {
      "calledWith": undefined,
      "name": "plugin:process|restart",
    },
  ],
]
`;

exports[`Update > updateAndRelaunch > Updates and relaunches > document.body.outerHTML 1`] = `"<body class="theme--system"></body>"`;

exports[`Update > updateAndRelaunch > Updates and relaunches > mockDb 1`] = `
{
  "deletedNotes": undefined,
  "encryptedNotes": [],
  "users": {
    "d": "1",
  },
}
`;

exports[`Update > updateAndRelaunch > Updates and relaunches > noteState 1`] = `
{
  "extraSelectedNotes": [],
  "notes": [],
  "selectedNote": {
    "content": {
      "body": "",
      "delta": {
        "ops": [],
      },
      "title": "",
    },
    "timestamp": 0,
    "uuid": "uuid",
  },
}
`;

exports[`Update > updateAndRelaunch > Updates and relaunches > openedPopup 1`] = `undefined`;

exports[`Update > updateAndRelaunch > Updates and relaunches > selectedTheme 1`] = `"System"`;

exports[`Update > updateAndRelaunch > Updates and relaunches > syncState 1`] = `
{
  "appError": AppError {
    "code": 0,
    "display": undefined,
    "message": undefined,
    "originalError": undefined,
    "retry": [Function],
    "retryConfig": undefined,
  },
  "encryptedNotesCache": Map {},
  "isLoggedIn": false,
  "loadingCount": 0,
  "newPassword": "",
  "password": "",
  "unsyncedNotes": UnsyncedNotesManager {
    "deleted": [],
    "edited": Set {},
    "new": "",
  },
  "username": "",
}
`;

exports[`Update > updateAndRelaunch > Updates and relaunches > updateState 1`] = `
{
  "isAvailable": false,
  "isDownloading": true,
  "strategy": "manual",
}
`;
